File: algorithms/lowest_common_ancestor.cpp
// 236. Lowest Common Ancestor of a Binary Tree
// Author: xianfeng.zhu@gmail.com

#include <stdio.h>
#include <vector>

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

// Recursive, DFS
class Solution
{
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q)
    {
        if (root == nullptr || root == p || root == q)
        {
            return root;
        }

        auto* lnode = lowestCommonAncestor(root->left, p, q);
        auto* rnode = lowestCommonAncestor(root->right, p, q);
        if (lnode != nullptr && rnode != nullptr)
        {
            return root;
        }
        return (lnode != nullptr ? lnode : rnode);
    }
};

int main(int argc, char* argv[])
{
    TreeNode* root = nullptr;
    TreeNode* p = nullptr;
    TreeNode* q = nullptr;
    auto* ancestor = Solution().lowestCommonAncestor(root, p, q);
    printf("Lowest common ancestor is %p\n", ancestor);
    return 0;
}

File: algorithms/lowest_common_ancestor.go
// 236. Lowest Common Ancestor of a Binary Tree
// Author: xianfeng.zhu@gmail.com
package main

import "fmt"

// Definition for TreeNode.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	node := root
	for node != nil {
		parentVal := node.Val
		if p.Val > parentVal && q.Val > parentVal {
			node = node.Right
		} else if p.Val < parentVal && q.Val < parentVal {
			node = node.Left
		} else {
			return node
		}
	}
	return nil
}

func main() {
	root := TreeNode{0, nil, nil}
	p := TreeNode{2, nil, nil}
	q := TreeNode{3, nil, nil}
	node := lowestCommonAncestor(&root, &p, &q)
	fmt.Printf("Output: %v\n", node)
}
